crates/engine/src/io.rs

```diff
@@ -38,15 +58,152 @@ pub trait OutputBuffers {
     fn stderr(&self) -> &[u8];
 }
 
+/// Wrapper around File w/ a convienient PathBuf for cloning
+pub struct PipeFile(pub File, pub PathBuf);
+
+impl PipeFile {
+    /// Constructs an instance from a set of PipeFile objects.
+    pub fn new(file: File, path: PathBuf) -> Self {
+        Self(file, path)
+    }
+}
+
+impl std::fmt::Debug for PipeFile {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PipeFile")
+            .field("File", &self.0)
+            .field("PathBuf", &self.1)
+            .finish()
+    }
+}
+
+impl Clone for PipeFile {
+    fn clone(&self) -> Self {
+        let f = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(&self.1)
+            .unwrap();
```

Comment:
Are you referring to the `clone` in the `PathBuf` or the overall `Clone` `impl` for `PipeFile`? How can I address your uneasiness?

Reply:
The overall Clone for PipeFile.  Because of the need to reopen the file, there's a risk of failure; because that is handled by `unwrap` (since Clone can't return a Result), any such failure turns into a panic, which doesn't feel like it should happen in a clone.

A possible way around it might be to have an `Arc<RwLock<File>>` instead of a plain `File` - then you are sharing a guarded reference to a shared file handle, rather than having to reopen from the path?

Reply:
Sounds good~ I've added this change, I'm just pushing now and will test it locally.

Reply:
Hmmm~ Adding the `Arc<RwLock<>>` actually broke it. Ironically, I think doing it this way might be breaking the pipe.

Reply:
Weird!  Well, if the unwrap is the only way it can work, and it's reliable in practice, then I guess we have to leave it for now.  Sorry for sending you down the rabbit hole.

Reply:
No worries~ I've actually learned a lot from our conversations! ;D

Reply:
It looks like that was really the issue~ It's now working on my end. With that said, I'll resolve this conversation ✅